src/main/java/net/snowflake/ingest/connection/JWTManager.java

```diff
@@ -0,0 +1,181 @@
+/*
+ * Copyright (c) 2012-2023 Snowflake Computing Inc. All rights reserved.
+ */
+
+package net.snowflake.ingest.connection;
+
+import com.nimbusds.jose.JOSEException;
+import com.nimbusds.jose.JWSAlgorithm;
+import com.nimbusds.jose.JWSHeader;
+import com.nimbusds.jose.JWSSigner;
+import com.nimbusds.jose.crypto.RSASSASigner;
+import com.nimbusds.jwt.JWTClaimsSet;
+import com.nimbusds.jwt.SignedJWT;
+import java.security.KeyPair;
+import java.util.Date;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+import net.snowflake.ingest.utils.Cryptor;
+
+/**
+ * This class manages creating and automatically renewing the JWT token
+ *
+ * @author obabarinsa
+ * @since 1.8
+ */
+final class JWTManager extends SecurityManager {
+  // the token lifetime is 59 minutes
+  private static final float LIFETIME_IN_MINUTES = 59;
+
+  // the renewal time is 54 minutes
+  private static final int RENEWAL_INTERVAL_IN_MINUTES = 54;
+
+  private static final String TOKEN_TYPE = "KEYPAIR_JWT";
+
+  // The public - private key pair we're using to connect to the service
+  private final transient KeyPair keyPair;
+
+  // the token itself
+  protected final AtomicReference<String> token;
+
+  /**
+   * Creates a JWTManager entity for a given account, user and KeyPair with a specified time to
+   * renew the token
+   *
+   * @param accountName - the snowflake account name of this user
+   * @param username - the snowflake username of the current user
+   * @param keyPair - the public/private key pair we're using to connect
+   * @param timeTillRenewal - the time measure until we renew the token
+   * @param unit the unit by which timeTillRenewal is measured
+   * @param telemetryService reference to the telemetry service
+   */
+  JWTManager(
+      String accountName,
+      String username,
+      KeyPair keyPair,
+      int timeTillRenewal,
+      TimeUnit unit,
+      TelemetryService telemetryService) {
+    super(accountName, username, telemetryService);
+    // if any of our arguments are null, throw an exception
+    if (keyPair == null) {
+      throw new IllegalArgumentException();
```

Comment:
FYI `Codecov` is reporting that this line is not covered by tests.

Reply:
Fixed.